src/allmydata/node.py

```diff
@@ -797,6 +827,38 @@ def _tub_portlocation(config):
     return tubport, location
 
 
+def set_tub_locations(i2p_provider, tor_provider, tub, portlocation):
+    """
+    Assign a Tub its listener locations.
+
+    :param i2p_provider: See ``allmydata.util.i2p_provider.create``.
+    :param tor_provider: See ``allmydata.util.tor_provider.create``.
+    """
+    if portlocation is None:
+        log.msg("Tub is not listening")
+    else:
+        tubport, location = portlocation
+        for port in tubport.split(","):
```

Comment:
just a comment of style.. the method extraction refactoring is great.

I might be wrong, so feel free to ignore this.


It looks like the important call from here is `tub.listenOn(port_or_endpoint)`
Maybe rename it to "_listenOn" and move the setLocation and log outside of this.

This is only setting/associating but also listening.

Then just return after tub not listening... you will then know that you don't need to read the rest of the code.

With an else... you will need to read all to code and see where else end and see if there is any extra code after else.

```suggestion
        log.msg("Tub is not listening")
        return

tubport, location = portlocation
for port in tubport.split(","):
```